### PR TITLE
ARROW-11889: [C++] Add parallelism to streaming CSV reader

### PR DESCRIPTION
**Pull request by westonpace**
This converts the parser & decoder into map functions and then creates the streaming CSV reader as an async generator.  Parallel readahead is then added on top of the parser/decoder to allow for parallel reads.

One thing that is lost at the moment is the ability to encounter a parsing error and then continue.  There was a python test that read in the first block, failed to convert the second block, and then successfully read in a third block.  I'm not sure if that restart behavior is important but if it is I can look into adding it.

Another thing that could be investigated in the future is combining the file readers and table readers more.  They already share some components but the parsing and decoding logic, while basically the same, is handled very differently.  The only real difference is that the table reader saves all the parsed blocks for re-parsing and the streaming reader does not.

### CODE REVIEW

**Comment by westonpace** [comment]
@pitrou @n3world This should probably wait until after ARROW-12996 is merged in.  I think it'd be easier to rebase ARROW-12996 into this PR than the other way around.

**Comment by westonpace** [comment]
I've rebased in the changes from #10509.  The behavior is only slightly different.  Opening the streaming CSV reader reads in the first record batch so the bytes_read will reflect that before any batch is read.  After that each time a batch is read in the next batch will be read in.  This means the read will not increment bytes_read.  If reading in parallel then bytes_read could potentially be even further ahead of the consumer since it will be doing decoding in readahead.  It should still match the spirit of the feature which is to report how many bytes have been decoded.

@n3world @pitrou review is welcome.  The CI failure is unrelated.

File: cpp/src/arrow/csv/reader.cc
Comment (n3world) [inline cpp/src/arrow/csv/reader.cc:514]: Should the `_` suffix be removed here like the other struct members?

File: cpp/src/arrow/csv/reader.cc
Comment (n3world) [inline cpp/src/arrow/csv/reader.cc:509]: This is only used by MakeColumnDecoders and probably could be passed in as an argument to just that method.

File: cpp/src/arrow/csv/reader.cc
Comment (n3world) [inline cpp/src/arrow/csv/reader.cc:722]: What is this class?

File: python/pyarrow/tests/test_csv.py
Comment (n3world) [inline python/pyarrow/tests/test_csv.py:1492]: Is this still to test SerialStreamingCSV? Should there be two classes so that all test get run for serial and non serial?

File: cpp/src/arrow/csv/reader_test.cc
Comment (westonpace) [reply cpp/src/arrow/csv/reader_test.cc]: I moved the increment of `decoded_bytes_` to be after the readahead.  So now...
 * bytes_decoded_ will not be incremented until the reader asks for the batch
 * The header bytes and skip before header are still marked read after Make (I think this is fair as they have been "consumed" by this point)
 * Bytes skipped after the header are marked consumed after the first batch is delivered

I think this is close enough to what you are after.

File: cpp/src/arrow/csv/reader.cc
Comment (westonpace) [reply cpp/src/arrow/csv/reader.cc]: Oops.  Removed.

File: cpp/src/arrow/csv/reader.cc
Comment (westonpace) [reply cpp/src/arrow/csv/reader.cc]: Thanks, added.

File: cpp/src/arrow/csv/reader.cc
Comment (westonpace) [reply cpp/src/arrow/csv/reader.cc]: I ended up getting rid of this member but yes, it should have been removed.

File: cpp/src/arrow/csv/reader.cc
Comment (westonpace) [reply cpp/src/arrow/csv/reader.cc]: Good observation, I've done this.

File: cpp/src/arrow/csv/reader.cc
Comment (westonpace) [reply cpp/src/arrow/csv/reader.cc]: Leftover from a previous implementation.  Removed.

File: python/pyarrow/tests/test_csv.py
Comment (westonpace) [reply python/pyarrow/tests/test_csv.py]: Ah, good point (and clumsy of me to leave those comments in there).  I've changed up the test so now there is no base class and every test is parameterized on use_threads=True/False.

**Comment by westonpace** [comment]
Thanks for the feedback @n3world.  I think I was able to update `bytes_read` to align with your use case a little better.

**Comment by westonpace** [comment]
@ursabot please benchmark

File: cpp/src/arrow/csv/reader.cc
Comment (bkietz) [inline cpp/src/arrow/csv/reader.cc:387]:
```suggestion
// A function object that takes in a buffer of CSV data and returns a parsed batch of CSV
// data (CSVBlock -> ParsedBlock) for use with MakeMappedGenerator.
// The parsed batch contains a list of offsets for each of the columns so that columns
// can be individually scanned
```

File: cpp/src/arrow/csv/reader.cc
Comment (bkietz) [inline cpp/src/arrow/csv/reader.cc:442]:
```suggestion
// A function object that takes in parsed batch of CSV data and decodes it to an arrow
// record batch (ParsedBlock -> DecodedBlock) for use with MakeMappedGenerator.
class BlockDecodingOperator {
```

File: cpp/src/arrow/csv/reader.cc
Comment (bkietz) [inline cpp/src/arrow/csv/reader.cc:451]:
```suggestion
    auto decoded_arrays_fut = All(std::move(decoded_array_futs));
```

File: cpp/src/arrow/csv/reader.cc
Comment (bkietz) [inline cpp/src/arrow/csv/reader.cc:400]:
```suggestion
    constexpr int32_t max_num_rows = std::numeric_limits<int32_t>::max();
```

File: cpp/src/arrow/csv/reader.cc
Comment (bkietz) [inline cpp/src/arrow/csv/reader.cc:487]:
```suggestion
        std::vector<std::shared_ptr<Array>> arrays) {
```

File: cpp/src/arrow/csv/reader.cc
Comment (bkietz) [inline cpp/src/arrow/csv/reader.cc:460]:
```suggestion
          ARROW_ASSIGN_OR_RAISE(auto batch,
                                state->DecodedArraysToBatch(std::move(decoded_arrays)));
```

File: cpp/src/arrow/csv/reader.cc
Comment (bkietz) [inline cpp/src/arrow/csv/reader.cc:882]: Nit: could you space these out a little?

File: cpp/src/arrow/csv/reader.cc
Comment (bkietz) [inline cpp/src/arrow/csv/reader.cc:1181]:
```suggestion
    // count_cb must return a value instead of Status/Future<> to work with
    // MakeMappedGenerator, and it must use a type with a valid end value to work with
```

**Review by bkietz — CHANGES_REQUESTED** [review]
Just a few comments. Overall this is a nice clarification of the streaming reader

**Review by bkietz — APPROVED** [review]
LGTM, thanks for doing this!

File: cpp/src/arrow/csv/reader.cc
Comment (westonpace) [reply cpp/src/arrow/csv/reader.cc:886]: Done.

**Pull request closed without merging**
